openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/strcompat/doc/man/man9/osfcompat.9.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) osfcompat.9.man,v 0.9.2.6 2006/04/05 09:43:34 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:43:34 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" osfcompat.9.man,v
.\" Revision 0.9.2.6  2006/04/05 09:43:34  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database strcompat.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so strcompat.macros
.\"
.\"
.TH OSFCOMPAT 9 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ DDI/DKI"
.\"
.\"
.SH NAME
.B osfcompat
\- OSF/1 \fISTREAMS\fR Compatibility
.\"
.\"
.SH SYNOPSIS
.PP
.B #define _OSF_SOURCE
.br
.B #include <sys/ddi.h>
.\"
.\"
.SH OVERVIEW
.PP
.B Linux Fast-STREAMS
provides a compatibility module,
.BR streams-osfcompat.o ,
that provides compatibility functions that smooth porting of drivers from the
.I Mac OpenTransport 1.5r2
.I STREAMS
environment to
.BR "Linux Fast-STREAMS" .
The functions provided by this compatibility module are described here.
.\"
.\"
.SH DESCRIPTION
.PP
Many of the functions that are provided by
.IR OSF/1 \(rg
are also provided by the core
.B streams.o
kernel module of
.BR "Linux Fast-STREAMS" .
Other
.IR OSF/1 \(rg-specific
functions are provided by the
.B streams-osfcompat.o
compatibility kernel module.
.\"
.\"
.SS "Core Functions"
.PP
Functions that are already available in
.B Linux Fast STREAMS
are as follows:
.HP
.BR adjmsg (9)\ \-\ \c
Removes the specified number of bytes from a message.
.PD 0
.HP
.BR allocb (9)\ \-\ \c
Allocates a message block.
.HP
.BR backq (9)\ \-\ \c
Gets a pointer to the previous queue.
.HP
.BR bcanput (9)\ \-\ \c
Tests for flow control in a specified priority band.
.HP
.BR bcmp (9)\ \-\ \c
Compares two byte strings.
.HP
.BR bcopy (9)\ \-\ \c
Copies a series of bytes with a specified limit.
.HP
.BR bzero (9)\ \-\ \c
Zeros a block of memory.
.HP
.BR bufcall (9)\ \-\ \c
Gets a buffer when allocb fails.
.HP
.BR canput (9)\ \-\ \c
Tests for room in a message queue.
.HP
.BR copyb (9)\ \-\ \c
Copies a message block.
.HP
.BR copyin (9)\ \-\ \c
Copies data from a user address space to a kernel address space.
.HP
.BR copyout (9)\ \-\ \c
Copies data from a kernel address space to a user address space.
.HP
.BR copymsg (9)\ \-\ \c
Copies a message to a new message.
.HP
.BR datamsg (9)\ \-\ \c
Tests whether a message is a data message.
.HP
.BR dupb (9)\ \-\ \c
Duplicates a message block descriptor.
.HP
.BR dupmsg (9)\ \-\ \c
Duplicates a message.
.HP
.BR enableok (9)\ \-\ \c
Enables a queue for service.
.HP
.BR esballoc (9)\ \-\ \c
Allocates a message block with a shared buffer.
.HP
.BR flushband (9)\ \-\ \c
Flushes messages for a specified priority band.
.HP
.BR flushq (9)\ \-\ \c
Removes a message from a queue.
.HP
.BR freeb (9)\ \-\ \c
Frees a message block.
.HP
.BR freemsg (9)\ \-\ \c
Frees all message blocks in a message.
.HP
.BR getq (9)\ \-\ \c
Gets a message from the front of the queue.
.HP
.BR insq (9)\ \-\ \c
Inserts a STREAMS message into a queue.
.HP
.BR linkb (9)\ \-\ \c
Concatenates two message blocks.
.HP
.BR major (9)\ \-\ \c
Returns the device major number.
.HP
.BR makedev (9)\ \-\ \c
Returns a dev_t.
.HP
.BR minor (9)\ \-\ \c
Returns the device minor number.
.HP
.BR msgdsize (9)\ \-\ \c
Returns the number of bytes in a message.
.HP
.BR noenable (9)\ \-\ \c
Prevents a queue from being scheduled.
.HP
.BR OTHERQ (9)\ \-\ \c
Gets a pointer to a module's other queue.
.HP
.BR pullupmsg (9)\ \-\ \c
Concatenates bytes in a message.
.HP
.BR putbq (9)\ \-\ \c
Places a message at the head of a queue.
.HP
.BR putctl (9)\ \-\ \c
Puts a control message on a queue.
.HP
.BR putctl1 (9)\ \-\ \c
Puts a control message on a queue.
.HP
.BR putnext (9)\ \-\ \c
Sends a message to the next module in the stream.
.HP
.BR putq (9)\ \-\ \c
Puts a message on a queue.
.HP
.BR qenable (9)\ \-\ \c
Enables a queue.
.HP
.BR qreply (9)\ \-\ \c
Sends a message in the reverse direction.
.HP
.BR qsize (9)\ \-\ \c
Finds the number of messages on a queue.
.HP
.BR RD (9)\ \-\ \c
Gets a pointer to a module's read queue.
.HP
.BR rmalloc (9)\ \-\ \c
Allocates size units from the given resource map.
.HP
.BR rmfree (9)\ \-\ \c
Frees space previously allocated into the specified resource map.
.HP
.BR rmget (9)\ \-\ \c
Allocates size units from the given resource map.
.HP
.BR rminit (9)\ \-\ \c
Initializes a resource map.
.HP
.BR rmvb (9)\ \-\ \c
Removes a message block from a message block.
.HP
.BR rmvq (9)\ \-\ \c
Removes a message block from a queue.
.HP
.BR sleep (9)\ \-\ \c
Puts a calling process to sleep.
.HP
.BR spl (9)\ \-\ \c
Sets the processor priority to mask different levels of interrupts.
.HP
.BR strlog (9)\ \-\ \c
Submits messages for logging.
.HP
.BR strqget (9)\ \-\ \c
Gets information about a queue.
.HP
.BR strqset (9)\ \-\ \c
Changes information about a queue.
.HP
.BR testb (9)\ \-\ \c
Checks for an available buffer.
.HP
.BR timeout (9)\ \-\ \c
Initializes a callout queue element.
.HP
.BR uiomove (9)\ \-\ \c
Moves data between user and system virtual space.
.HP
.BR unbufcall (9)\ \-\ \c
Cancels a pending bufcall request.
.HP
.BR unlinkb (9)\ \-\ \c
Removes a message block from the head of a message.
.HP
.BR untimeout (9)\ \-\ \c
Removes the scheduled interface from the callout queues.
.HP
.BR vtop (9)\ \-\ \c
Converts any virtual address to a physical address.
.HP
.BR wakeup (9)\ \-\ \c
Wakes up all processes sleeping on a specified address.
.HP
.BR WR (9)\ \-\ \c
Gets a pointer to this module's write queue.
.PD
.\"
.SS "Specific Functions"
.PP
Functions that are available in the
.B Linux Fast-STREAMS
.IR OSF/1 \(rg
compatibility module,
.BR streams-osfcompat.o ,
are as follows:
.HP
.BR DELAY (9)\ \-\ \c
Delays the calling interface a specified number of microseconds.
.PD 0
.HP
.BR lbolt (9)\ \-\ \c
.HP
.BR MALLOC (9)\ \-\ \c
Allocates a variable-size section of kernel virtual memory.
.HP
.BR printf (9)\ \-\ \c
Prints text to the console and the error logger.
.HP
.BR puthere (9)\ \-\ \c
.HP
.BR streams_close_comm (9)\ \-\ \c
.HP
.BR streams_open_comm (9)\ \-\ \c
.HP
.BR streams_open_ocomm (9)\ \-\ \c
.HP
.BR time (9)\ \-\ \c
.HP
.BR uprintf (9)\ \-\ \c Is a nonsleeping kernel printf function.
.PD
.\"
.SS "Transport Functions"
.PP
.PD 0
.PD
.\"
.\"
.SH USAGE
.PP
.\"
.\"
.SH RETURN
.PP
.\"
.\"
.SH ERRORS
.PP
.\"
.\"
.SH CONTEXT
.PP
.\"
.\"
.SH MP-STREAMS
.PP
.\"
.\"
.SH NOTICES
.PP
These are
.IR OSF/1 \(rg-specific
functions:
.B _OSF_SOURCE
must be defined before including DDI/DKI include files to expose these symbols.
.B Linux Fast-STREAMS
must also have been compiled with
.B CONFIG_STREAMS_COMPAT_OSF
or
.B CONFIG_STREAMS_COMPAT_OSF_MODULE
defined.
.\"
.\"
.SH IMPLEMENTATION
.PP
.\"
.\"
.SH IOCTLS
.PP
.\"
.\"
.SH SYSCTLS
.PP
.\"
.\"
.SH "SEE ALSO"
.PP
Core functions:
.BR adjmsg (9),
.PD 0
.BR allocb (9),
.BR backq (9),
.BR bcanput (9),
.BR bcmp (9),
.BR bcopy (9),
.BR bzero (9),
.BR bufcall (9),
.BR canput (9),
.BR copyb (9),
.BR copyin (9),
.BR copyout (9),
.BR copymsg (9),
.BR datamsg (9),
.BR dupb (9),
.BR dupmsg (9),
.BR enableok (9),
.BR esballoc (9),
.BR flushband (9),
.BR flushq (9),
.BR freeb (9),
.BR freemsg (9),
.BR getq (9),
.BR insq (9),
.BR linkb (9),
.BR major (9),
.BR makedev (9),
.BR minor (9),
.BR msgdsize (9),
.BR noenable (9),
.BR OTHERQ (9),
.BR pullupmsg (9),
.BR putbq (9),
.BR putctl (9),
.BR putctl1 (9),
.BR putnext (9),
.BR putq (9),
.BR qenable (9),
.BR qreply (9),
.BR qsize (9),
.BR RD (9),
.BR rmalloc (9),
.BR rmfree (9),
.BR rmget (9),
.BR rminit (9),
.BR rmvb (9),
.BR rmvq (9),
.BR sleep (9),
.BR spl (9),
.BR strlog (9),
.BR strqget (9),
.BR strqset (9),
.BR testb (9),
.BR timeout (9),
.BR uiomove (9),
.BR unbufcall (9),
.BR unlinkb (9),
.BR untimeout (9),
.BR vtop (9),
.BR wakeup (9),
.BR WR (9).
.PP
Compatibility functions:
.BR DELAY (9),
.BR MALLOC (9),
.BR printf (9),
.BR uprintf (9).
.PP
Transport functions:
.PP
General:
.BR STREAMS (9).
.\"
.\"
.SH BUGS
.PP
.\"
.\"
.SH COMPATIBILITY
.PP
These functions are compatible with
.IR OSF/1 \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.IP \(em
.IP \(em
These are
.IR OSF/1 \(rg-specific
functions and should not be used by portable
.I STREAMS
drivers and modules.
The more common
.I SVR 4.2 MP
functions should be used in their stead.
.IP \(em
These functions
are provided for source level compatibility with
.IR OSF/1 \(rg
for the purpose of porting
.IR OSF/1 \(rg
drivers to
.BR Linux .
Because internal structures and definitions for
.B Linux Fast-STREAMS
.[
lfs
.]
differ from that of
.IR OSF/1 \(rg,
binaries are incompatible.
.IP \(em
These functions
only provide functionality comparable to
.IR OSF/1 \(rg
for use by
.I STREAMS
drivers and modules.
Any and all functionality provided by
these functions
in support of
.RI non- STREAMS
devices are not supported.
.PP
See
.BR STREAMS (9)
for additional compatibility considerations.
.\"
.\"
.SH CONFORMANCE
.PP
.IR OSF/1 \(rg
DDI documentation.
.[
osfdoc
.]
.[
osfspg
.]
.\"
.\"
.SH HISTORY
.PP
Most of these functions
first appeared in
.IR OSF/1 \(rg.
Functions with name starting with
.I mi_
are probably Mentat, Inc. functions that goe all the way back to the
.RI pre- OSF/1
.I Mentat
implementations.
.\"
.\"
.[
$LIST$
.]
.TI


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/strcompat/src/kernel/osfcompat.c
/*****************************************************************************

 @(#) osfcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.19) 2006/07/07 20:49:55

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2005  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2006/07/07 20:49:55 by brian

 *****************************************************************************/

#ident "@(#) osfcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.19) 2006/07/07 20:49:55"

static char const ident[] =
    "osfcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.19) 2006/07/07 20:49:55";

/* 
 *  This is my solution for those who don't want to inline GPL'ed functions or
 *  who don't use optimizations when compiling or specifies
 *  -fnoinline-functions or something of the like.  This file implements all
 *  of the extern inlines from the header files by just including the header
 *  files with the functions declared 'inline' instead of 'extern inline'.
 *
 *  There are implemented here in a separate object, out of the way of the
 *  modules that don't use them.
 */

#define __OSF_EXTERN_INLINE __inline__ streamscall

#define _OSF_SOURCE

#include "sys/os7/compat.h"

#define OSFCOMP_DESCRIP		"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define OSFCOMP_COPYRIGHT	"Copyright (c) 1997-2005 OpenSS7 Corporation.  All Rights Reserved."
#define OSFCOMP_REVISION	"LfS osfcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.19) 2006/07/07 20:49:55"
#define OSFCOMP_DEVICE		"OSF/1.2 Compatibility"
#define OSFCOMP_CONTACT		"Brian Bidulock <bidulock@openss7.org>"
#define OSFCOMP_LICENSE		"GPL"
#define OSFCOMP_BANNER		OSFCOMP_DESCRIP		"\n" \
				OSFCOMP_COPYRIGHT	"\n" \
				OSFCOMP_REVISION	"\n" \
				OSFCOMP_DEVICE		"\n" \
				OSFCOMP_CONTACT		"\n"
#define OSFCOMP_SPLASH		OSFCOMP_DEVICE		" - " \
				OSFCOMP_REVISION	"\n"

#ifdef CONFIG_STREAMS_COMPAT_OSF_MODULE
MODULE_AUTHOR(OSFCOMP_CONTACT);
MODULE_DESCRIPTION(OSFCOMP_DESCRIP);
MODULE_SUPPORTED_DEVICE(OSFCOMP_DEVICE);
MODULE_LICENSE(OSFCOMP_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-osfcompat");
#endif
#endif

__OSF_EXTERN_INLINE void puthere(queue_t *q, mblk_t *mp);

EXPORT_SYMBOL_NOVERS(puthere);
__OSF_EXTERN_INLINE time_t lbolt(void);

EXPORT_SYMBOL_NOVERS(lbolt);
__OSF_EXTERN_INLINE time_t time(void);

EXPORT_SYMBOL_NOVERS(time);

struct str_comm {
	struct str_comm **prev;		/* must be first */
	struct str_comm *next;
	dev_t dev;			/* device number (or NODEV for modules) */
	size_t size;			/* size of private data */
	char priv[0];			/* followed by private data */
};

static spinlock_t str_list_lock = SPIN_LOCK_UNLOCKED;
static struct str_comm *str_list = NULL;

/* 
 *  STREAMS_OPEN_COMM
 *  -------------------------------------------------------------------------
 *  for V4 open
 */
int
streams_open_comm(unsigned int size, queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	struct str_comm *sp, **spp = &str_list;
	major_t cmajor = getmajor(*devp);
	minor_t cminor = getminor(*devp);

	if (q->q_ptr != NULL)
		return (0);	/* already open */
	if (sflag == MODOPEN) {
		if (!WR(q)->q_next)
			return (EIO);
	} else {
		if (WR(q)->q_next && SAMESTR(q))
			return (EIO);
	}
	if ((sp = kmem_alloc(sizeof(*sp) + size, KM_NOSLEEP)))	/* we could probably sleep */
		return (ENOMEM);
	bzero(sp, sizeof(*sp) + size);
	spin_lock(&str_list_lock);
	switch (sflag) {
	case CLONEOPEN:
		/* first clone minor (above 5 per AIX docs), but the caller can start wherever they 
		   want above that */
		if (cminor <= 5)
			cminor = 5 + 1;
		/* fall through */
	default:
	case DRVOPEN:
	{
		major_t dmajor = cmajor;

		for (; *(spp) && (dmajor = getmajor((*spp)->dev)) < cmajor; spp = &(*spp)->next) ;
		for (; *(spp) && dmajor == getmajor((*spp)->dev) &&
		     getminor(makedevice(0, cminor)) != 0; spp = &(*spp)->next, cminor++) {
			minor_t dminor = getminor((*spp)->dev);

			if (cminor < dminor)
				break;
			if (cminor == dminor)
				if (sflag != CLONEOPEN) {
					spin_unlock(&str_list_lock);
					kmem_free(sp, sizeof(*sp) + size);
					return (ENXIO);
				}
		}
		if (getminor(makedevice(0, cminor)) == 0) {	/* no minors left */
			spin_unlock(&str_list_lock);
			kmem_free(sp, sizeof(*sp) + size);
			return (EBUSY);
		}
		sp->dev = makedevice(cmajor, cminor);
		break;
	}
	case MODOPEN:
	{
		/* just push modules on list with no device */
#ifdef NODEV
		sp->dev = NODEV;
#else
		sp->dev = 0;
#endif
		break;
	}
	}
	sp->size = size;
	if ((sp->next = *spp))
		sp->next->prev = &sp;
	sp->prev = spp;
	*spp = sp;
	q->q_ptr = OTHERQ(q)->q_ptr = (sp + 1);
	spin_unlock(&str_list_lock);
	return (0);
}

EXPORT_SYMBOL_NOVERS(streams_open_comm);

/* 
 *  STREAMS_OPEN_OCOMM
 *  -------------------------------------------------------------------------
 *  for V3 open
 */
int
streams_open_ocomm(dev_t dev, unsigned int size, queue_t *q, dev_t *devp, int oflag, int sflag,
		   cred_t *crp)
{
	dev_t mydev = dev;
	int err;

	if ((err = streams_open_comm(size, q, &mydev, oflag, sflag, crp)))
		return (err);
	*devp = mydev;
	return (0);
}

EXPORT_SYMBOL_NOVERS(streams_open_ocomm);

/* 
 *  STREAMS_CLOSE_COMM
 *  -------------------------------------------------------------------------
 *  for both V3 and V4 close
 */
int
streams_close_comm(queue_t *q, int oflag, cred_t *crp)
{
	spin_lock(&str_list_lock);
	if (q->q_ptr) {
		struct str_comm *sp = ((struct str_comm *) q->q_ptr) - 1;
		size_t size = sp->size;

		/* found it */
		if ((*sp->prev = sp->next))
			sp->next->prev = sp->prev;
		sp->next = NULL;
		sp->prev = &sp->next;
		sp->size = 0;
		q->q_ptr = OTHERQ(q)->q_ptr = NULL;
		(void) size;	/* LiS doesn't actually use this */
		kmem_free(sp, sizeof(*sp) + size);
	}
	spin_unlock(&str_list_lock);
	return (0);
}

EXPORT_SYMBOL_NOVERS(streams_close_comm);

#ifndef NODEV
#define NODEV (makedevice(0, 0))
#endif
/*
 *  STRMOD_ADD
 *  -------------------------------------------------------------------------
 */
dev_t
strmod_add(dev_t dev, struct streamtab *st, struct streamadm *sa)
{
#ifdef LIS
	switch (sa->sa_flags & STR_TYPE_MASK) {
	case STR_IS_DEVICE:
	{
		major_t major = getmajor(dev);
		minor_t minor = getminor(dev);
		int result;

		if ((result = lis_register_strdev(major, st, minor, sa->sa_name)) > 0) {
			major = result;
			dev = makedevice(major, minor);
		} else {
			dev = NODEV;
		}
		return (dev);
	}
	case STR_IS_MODULE:
	{
		modID_t modid;
		int result;

		if ((result = lis_register_strmod(st, sa->sa_name)) > 0) {
			modid = result;
			dev = makedevice(modid, 0);
		} else {
			dev = NODEV;
		}
		return (dev);
	}
	default:
		return (NODEV);
	}
#endif
#ifdef LFS
	switch (sa->sa_flags & STR_TYPE_MASK) {
	case STR_IS_DEVICE:
	{
		struct cdevsw *cdev;
		int err;

		if (!(cdev = kmem_zalloc(sizeof(*cdev), KM_NOSLEEP)))
			return (NODEV);
		cdev->d_name = sa->sa_name;
		cdev->d_str = st;
		/* build flags */
		cdev->d_flag = 0;
		if (sa->sa_flags & STR_QSAFETY) {
			cdev->d_flag |= D_SAFE;
		}
		switch ((cdev->d_sqlvl = sa->sa_sync_level)) {
		case SQLVL_NOP:
			cdev->d_flag |= D_MP;
			break;
		case SQLVL_QUEUE:
			cdev->d_flag |= D_MTPERQ;
			break;
		case SQLVL_QUEUEPAIR:
			cdev->d_flag |= D_MTQPAIR;
			break;
		case SQLVL_MODULE:
			cdev->d_flag |= D_MTPERMOD;
			break;
		case SQLVL_ELSEWHERE:
			cdev->d_flag |= D_MTOUTPERIM;
			break;
		case SQLVL_GLOBAL:
			/* can't really support this, but its only used for debug anyway */
			cdev->d_flag &= ~D_MP;
			break;
		case SQLVL_DEFAULT:
			cdev->d_flag |= D_MTPERMOD;
			break;
		}
		if ((err = register_strdev(cdev, getmajor(dev))) < 0) {
			kmem_free(cdev, sizeof(*cdev));
			return (NODEV);
		}
		return (makedevice(err, getminor(dev)));
	}
	case STR_IS_MODULE:
	{
		struct fmodsw *fmod;
		int err;

		if (!(fmod = kmem_zalloc(sizeof(*fmod), KM_NOSLEEP)))
			return (NODEV);
		fmod->f_name = sa->sa_name;
		fmod->f_str = st;
		/* build flags */
		fmod->f_flag = 0;
		if (sa->sa_flags & STR_QSAFETY) {
			fmod->f_flag |= D_SAFE;
		}
		switch ((fmod->f_sqlvl = sa->sa_sync_level)) {
		case SQLVL_NOP:
			fmod->f_flag |= D_MP;
			break;
		case SQLVL_QUEUE:
			fmod->f_flag |= D_MTPERQ;
			break;
		case SQLVL_QUEUEPAIR:
			fmod->f_flag |= D_MTQPAIR;
			break;
		case SQLVL_MODULE:
			fmod->f_flag |= D_MTPERMOD;
			break;
		case SQLVL_ELSEWHERE:
			fmod->f_flag |= D_MTOUTPERIM;
			break;
		case SQLVL_GLOBAL:
			/* can't really support this, but its only used for debug anyway */
			fmod->f_flag &= ~D_MP;
			break;
		case SQLVL_DEFAULT:
			fmod->f_flag |= D_MTPERMOD;
			break;
		}
		if ((err = register_strmod(fmod)) < 0) {
			kmem_free(fmod, sizeof(*fmod));
			return (NODEV);
		}
		return (makedevice(err, 0));
	}
	default:
		return (NODEV);
	}
#endif
}

EXPORT_SYMBOL_NOVERS(strmod_add);

/*
 *  STRMOD_DEL
 *  -------------------------------------------------------------------------
 */
int
strmod_del(dev_t dev, struct streamtab *st, struct streamadm *sa)
{
#ifdef LIS
	switch (sa->sa_flags & STR_TYPE_MASK) {
	case STR_IS_DEVICE:
		return lis_unregister_strdev(getmajor(dev));
	case STR_IS_MODULE:
		return lis_unregister_strmod(st);
	default:
		return (EINVAL);
	}
#endif
#ifdef LFS
	switch (sa->sa_flags & STR_TYPE_MASK) {
	case STR_IS_DEVICE:
	{
		struct cdevsw *cdev;
		int err;

		if ((cdev = cdev_str(st)) == NULL)
			return (ENOENT);
		if ((err = unregister_strdev(cdev, getmajor(dev))) == 0)
			kmem_free(cdev, sizeof(*cdev));
		return (-err);
	}
	case STR_IS_MODULE:
	{
		struct fmodsw *fmod;
		int err;

		if (!(fmod = fmod_str(st)))
			return (ENOENT);
		if ((err = unregister_strmod(fmod)) == 0)
			kmem_free(fmod, sizeof(fmod));
		return (-err);
	}
	default:
		return (EINVAL);
	}
#endif
}

EXPORT_SYMBOL_NOVERS(strmod_del);

#ifdef CONFIG_STREAMS_COMPAT_OSF_MODULE
static
#endif
int __init
osfcomp_init(void)
{
#ifdef CONFIG_STREAMS_COMPAT_OSF_MODULE
	printk(KERN_INFO OSFCOMP_BANNER);
#else
	printk(KERN_INFO OSFCOMP_SPLASH);
#endif
	return (0);
}

#ifdef CONFIG_STREAMS_COMPAT_OSF_MODULE
static
#endif
void __exit
osfcomp_exit(void)
{
	return;
}

#ifdef CONFIG_STREAMS_COMPAT_OSF_MODULE
module_init(osfcomp_init);
module_exit(osfcomp_exit);
#endif
